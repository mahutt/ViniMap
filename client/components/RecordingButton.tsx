import React, { useCallback, useState } from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import RNUxcam from 'react-native-ux-cam';

interface ActionButtonProps {
  onPress: () => void;
  buttonStyle: object;
  label: string;
}

const ActionButton: React.FC<ActionButtonProps> = ({ onPress, buttonStyle, label }) => (
  <TouchableOpacity style={[styles.button, buttonStyle]} onPress={onPress}>
    <Text style={styles.buttonText}>{label}</Text>
  </TouchableOpacity>
);

const RecordingButton: React.FC = () => {
  const [isRecording, setIsRecording] = useState(false);

  const startRecording = useCallback(() => {
    console.log('Starting new UXCam session...');
    RNUxcam.startNewSession();
    console.log('UXCam session started');
    setIsRecording(true);
  }, [setIsRecording]);

  const stopRecording = useCallback(() => {
    console.log('Stopping UXCam session...');
    RNUxcam.stopSessionAndUploadData();
    console.log('UXCam session stopped and upload requested');
    setIsRecording(false);
  }, [setIsRecording]);

  return isRecording ? (
    <ActionButton
      onPress={stopRecording}
      buttonStyle={styles.stopButton}
      label="Stop Recording (Usability test)"
    />
  ) : (
    <ActionButton
      onPress={startRecording}
      buttonStyle={styles.startButton}
      label="Start Recording (Usability test)"
    />
  );
};

const styles = StyleSheet.create({
  button: {
    padding: 12,
    borderRadius: 8,
    marginTop: 16,
  },
  buttonText: {
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
  startButton: {
    backgroundColor: '#007AFF',
  },
  stopButton: {
    backgroundColor: '#FF3B30',
  },
});

export default RecordingButton;
