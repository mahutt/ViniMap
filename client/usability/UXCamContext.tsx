import React, {
  createContext,
  ReactNode,
  useEffect,
  useState,
  useCallback,
  useContext,
  useMemo,
} from 'react';
import { usePathname } from 'expo-router';
import RNUxcam from 'react-native-ux-cam';

interface UXCamContextType {
  isRecording: boolean;
  startRecording: () => void;
  stopRecording: () => void;
}

const UXCamContext = createContext<UXCamContextType | null>(null);

interface UXCamProviderProps {
  children: ReactNode;
}

export function UXCamProvider({ children }: UXCamProviderProps): JSX.Element {
  const pathname = usePathname();
  const [isRecording, setIsRecording] = useState(false);

  useEffect(() => {
    const uxcam_api_key = process.env.EXPO_PUBLIC_UXCAM_API_KEY;

    if (uxcam_api_key) {
      RNUxcam.optIntoSchematicRecordings(); // enable ios screen recordings

      const configuration = {
        userAppKey: uxcam_api_key,
        enableAutomaticScreenNameTagging: false,
        enableAdvancedGestureRecognition: true,
        enableImprovedScreenCapture: true,
        disableAutoRecord: true,
        occlusionType: 0,
      };

      RNUxcam.startWithConfiguration(configuration);
    } else {
      console.warn('api key for UXCam undefined');
    }
  }, []);

  // handling screen name changes for usability testing
  useEffect(() => {
    if (pathname) {
      let screenName = 'Unknown';

      if (pathname === '/') {
        screenName = 'HomeScreen';
      } else if (pathname.startsWith('/')) {
        screenName = pathname.substring(1).replace(/\//g, '_').replace(/\W/g, '');

        screenName = screenName
          .split('_')
          .map((segment) => segment.charAt(0).toUpperCase() + segment.slice(1))
          .join('_');

        if (screenName.length === 0) {
          screenName = 'HomeScreen';
        } else {
          screenName += 'Screen';
        }
      }
      RNUxcam.tagScreenName(screenName);
    }
  }, [pathname]);

  const startRecording = useCallback(() => {
    console.log('Starting new UXCam session...');
    RNUxcam.startNewSession();
    console.log('UXCam session started');
    setIsRecording(true);
  }, []);

  const stopRecording = useCallback(() => {
    console.log('Stopping UXCam session...');
    RNUxcam.stopSessionAndUploadData();
    console.log('UXCam session stopped and upload requested');
    setIsRecording(false);
  }, []);

  const UXCamContextValue = useMemo(
    () => ({
      isRecording,
      startRecording,
      stopRecording,
    }),
    [isRecording, startRecording, stopRecording]
  );

  return <UXCamContext.Provider value={UXCamContextValue}>{children}</UXCamContext.Provider>;
}

export function useUXCam() {
  const context = useContext(UXCamContext);
  if (context === null) {
    throw new Error('useUXCam must be used within an UXCamProvider');
  }
  return context;
}
