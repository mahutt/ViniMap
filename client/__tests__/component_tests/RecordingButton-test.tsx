import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import RecordingButton from '@/components/RecordingButton';
import { useUXCam } from '@/usability/UXCamContext';

// Mock the UXCamContext hook
jest.mock('@/usability/UXCamContext', () => ({
  useUXCam: jest.fn(),
}));

const mockedUseUXCam = useUXCam as jest.MockedFunction<typeof useUXCam>;

describe('RecordingButton Component', () => {
  describe('ActionButton', () => {
    it('renders correctly with given props', () => {
      mockedUseUXCam.mockReturnValue({
        isRecording: false,
        startRecording: jest.fn(),
        stopRecording: jest.fn(),
      });

      const { getByText } = render(<RecordingButton />);

      const button = getByText('Start Recording (Usability test)');
      expect(button).toBeTruthy();
    });
  });

  // Test for RecordingButton when not recording
  describe('RecordingButton when not recording', () => {
    beforeEach(() => {
      // Mock the UXCam hook for the not recording state
      const startRecording = jest.fn();
      mockedUseUXCam.mockReturnValue({
        isRecording: false,
        startRecording,
        stopRecording: jest.fn(),
      });
    });

    it('renders the Start Recording button when not recording', () => {
      const { getByText } = render(<RecordingButton />);
      expect(getByText('Start Recording (Usability test)')).toBeTruthy();
    });

    it('calls startRecording when Start Recording button is pressed', () => {
      const { getByText } = render(<RecordingButton />);
      const startButton = getByText('Start Recording (Usability test)');

      fireEvent.press(startButton);

      const { startRecording } = mockedUseUXCam();
      expect(startRecording).toHaveBeenCalledTimes(1);
    });
  });

  // Test for RecordingButton when recording
  describe('RecordingButton when recording', () => {
    beforeEach(() => {
      // Mock the UXCam hook for the recording state
      const stopRecording = jest.fn();
      mockedUseUXCam.mockReturnValue({
        isRecording: true,
        startRecording: jest.fn(),
        stopRecording,
      });
    });

    it('renders the Stop Recording button when recording', () => {
      const { getByText } = render(<RecordingButton />);
      expect(getByText('Stop Recording (Usability test)')).toBeTruthy();
    });

    it('calls stopRecording when Stop Recording button is pressed', () => {
      const { getByText } = render(<RecordingButton />);
      const stopButton = getByText('Stop Recording (Usability test)');

      fireEvent.press(stopButton);

      const { stopRecording } = mockedUseUXCam();
      expect(stopRecording).toHaveBeenCalledTimes(1);
    });
  });

  // Test for styling
  describe('Component styling', () => {
    it('applies the correct styles for Start Recording button', () => {
      mockedUseUXCam.mockReturnValue({
        isRecording: false,
        startRecording: jest.fn(),
        stopRecording: jest.fn(),
      });

      const { getByText } = render(<RecordingButton />);
      const startButton = getByText('Start Recording (Usability test)');

      expect(startButton).toBeTruthy();
    });

    it('applies the correct styles for Stop Recording button', () => {
      mockedUseUXCam.mockReturnValue({
        isRecording: true,
        startRecording: jest.fn(),
        stopRecording: jest.fn(),
      });

      const { getByText } = render(<RecordingButton />);
      const stopButton = getByText('Stop Recording (Usability test)');

      expect(stopButton).toBeTruthy();
    });
  });

  // Test for snapshot testing
  it('matches snapshot when not recording', () => {
    mockedUseUXCam.mockReturnValue({
      isRecording: false,
      startRecording: jest.fn(),
      stopRecording: jest.fn(),
    });

    const { toJSON } = render(<RecordingButton />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when recording', () => {
    mockedUseUXCam.mockReturnValue({
      isRecording: true,
      startRecording: jest.fn(),
      stopRecording: jest.fn(),
    });

    const { toJSON } = render(<RecordingButton />);
    expect(toJSON()).toMatchSnapshot();
  });
});
