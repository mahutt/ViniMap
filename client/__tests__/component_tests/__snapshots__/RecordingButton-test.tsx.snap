// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordingButton Component matches snapshot when not recording 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#007AFF",
      "borderRadius": 8,
      "marginTop": 16,
      "opacity": 1,
      "padding": 12,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    Start Recording (Usability test)
  </Text>
</View>
`;

exports[`RecordingButton Component matches snapshot when recording 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FF3B30",
      "borderRadius": 8,
      "marginTop": 16,
      "opacity": 1,
      "padding": 12,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    Stop Recording (Usability test)
  </Text>
</View>
`;
